File: bst.js
class Node {
  constructor(val) {
    this.val = val;
    this.right = null;
    this.left = null;
  }
}

class BinarySearchTree {
  constructor(root = null) {
    this.root = root;
  }

  find(rootNode, val) {
    if (rootNode.val === val) {
      return true;
    }
    if (!this.root) {
      return false;
    } else {
      if (rootNode.val > val && rootNode.left) {
        return this.find(rootNode.left, val);
      } else if (rootNode.val < val && rootNode.right) {
        return this.find(rootNode.right, val);
      } else {
        return false;
      }
    }
  }

  insert(val) {
    let node = new Node(val);
    if (!this.root) {
      this.root = node;
    } else {
      this.insertNode(this.root, node);
    }
  }

  insertNode(node, newNode) {
    if (node.val > newNode.val) {
      if (!node.left) {
        node.left = newNode;
      } else {
        this.insertNode(node.left, newNode);
      }
    } else {
      if (!node.right) {
        node.right = newNode;
      } else {
        this.insertNode(node.right, newNode);
      }
    }
  }
}

let bst = new BinarySearchTree();
bst.insert(10);
bst.insert(20);
bst.insert(8);
bst.insert(9);
bst.insert(25);
bst.insert(4);
bst.insert(18);

const BFS = root => {
  let visited = [];
  let q = [root];

  // key is to understand that you can traverse what gets added into the QUEUE vs thinking recursively

  while (q.length > 0) {
    // ref to the curr top elem in q
    let deqNode = q.shift();
    // push deqNode into visited
    visited.push(deqNode.val);
    // check if deqNode has right or left children
    if (deqNode.left || deqNode.right) {
      q.push(deqNode.left);
      q.push(deqNode.right);
    }
  }

  return visited;
};

// console.log(BFS(bst.root));

const DFS = root => {
  let visited = [];

  if (!root) {
    return;
  } else {
    if (root.left) {
      visited = visited.concat(DFS(root.left));
    }
    visited.push(root.val);
    if (root.right) {
      visited = visited.concat(DFS(root.right));
    }
  }
  return visited;
};

console.log(DFS(bst.root));

File: bst.py

# create the node
class Node:
    def __init__(self, value):
        self.value = value
        self.right = None
        self.left = None
    
    def __repr__(self):
        return "Node: %s" % self.value
    # def __str__(self):
    #     return str(self.__class__) + ": " + str(self.__dict__)

class BST:
    def __init__(self, root):
        self.root = root
        self.size = 1

    def __str__(self):
        return "BST " + str(self.__dict__)

    def insert(self, node):
        if(node is None):
            return 'Insert a valid node'
        
        # if there's right children or left children then must check accordingly starting from root
        if(self.root.value < node.value):
            # now check for children
            if (self.root.right is None):
                self.root.right = BST(node)
            else:
                self.root.right.insert(node)
        else:
            if (self.root.left is None):
                self.root.left = BST(node)
            else:
                self.root.left.insert(node)
        
        self.size += 1
        return node



def maxDepthGivenNode(bstNode):
    depth = 0
    # if there's no left or right child then return 0 count
    if(bstNode is None or (bstNode.root.left is None and bstNode.root.right is None)):
        return depth

    # no traverse left or right to find depth
    depth += 1
    leftDepth = maxDepthGivenNode(bstNode.root.left)
    rightDepth = maxDepthGivenNode(bstNode.root.right)

    if(leftDepth > rightDepth):
        depth += leftDepth
    else:
        depth += rightDepth
    
    return depth
        
root = Node(5)
node2 = Node(4)
node3 = Node(2)
node4 = Node(1)
print(root)

newBST = BST(root)
newBST.insert(node2)
newBST.insert(node3)
newBST.insert(node4)

print(newBST)
# print(maxDepthGivenNode(newBST.root.left))
